File: svws-webclient/core/src/core/utils/klausurplanung/KlausurterminblockungAlgorithmus.ts
import { JavaObject } from '../../../java/lang/JavaObject';
import { GostKlausurterminblockungErgebnis } from '../../../core/data/gost/klausurplanung/GostKlausurterminblockungErgebnis';
import { KlausurterminblockungModusKursarten } from '../../../core/types/gost/klausurplanung/KlausurterminblockungModusKursarten';
import { GostKlausurterminblockungDaten } from '../../../core/data/gost/klausurplanung/GostKlausurterminblockungDaten';
import { GostKursklausur } from '../../../core/data/gost/klausurplanung/GostKursklausur';
import { GostKlausurterminblockungKonfiguration } from '../../../core/data/gost/klausurplanung/GostKlausurterminblockungKonfiguration';
import { ArrayList } from '../../../java/util/ArrayList';
import { KlausurterminblockungAlgorithmusGreedy1 } from '../../../core/utils/klausurplanung/KlausurterminblockungAlgorithmusGreedy1';
import { KlausurterminblockungAlgorithmusGreedy2 } from '../../../core/utils/klausurplanung/KlausurterminblockungAlgorithmusGreedy2';
import { DeveloperNotificationException } from '../../../core/exceptions/DeveloperNotificationException';
import { KlausurterminblockungAlgorithmusGreedy3 } from '../../../core/utils/klausurplanung/KlausurterminblockungAlgorithmusGreedy3';
import { Logger, cast_de_svws_nrw_core_logger_Logger } from '../../../core/logger/Logger';
import { System } from '../../../java/lang/System';
import type { Comparator } from '../../../java/util/Comparator';
import { Random } from '../../../java/util/Random';
import { KlausurterminblockungDynDaten } from '../../../core/utils/klausurplanung/KlausurterminblockungDynDaten';
import { KlausurterminblockungAlgorithmusGreedy1b } from '../../../core/utils/klausurplanung/KlausurterminblockungAlgorithmusGreedy1b';
import { KlausurterminblockungModusQuartale } from '../../../core/types/gost/klausurplanung/KlausurterminblockungModusQuartale';
import { KlausurterminblockungAlgorithmusAbstract } from '../../../core/utils/klausurplanung/KlausurterminblockungAlgorithmusAbstract';
import { KlausurterminblockungAlgorithmusGreedy2b } from '../../../core/utils/klausurplanung/KlausurterminblockungAlgorithmusGreedy2b';
import type { List } from '../../../java/util/List';

export class KlausurterminblockungAlgorithmus extends JavaObject {

	private static readonly _random : Random = new Random();

	private static readonly _compGostKursklausur : Comparator<GostKursklausur> = { compare : (a: GostKursklausur, b: GostKursklausur) => {
		if (a.halbjahr < b.halbjahr)
			return -1;
		if (a.halbjahr > b.halbjahr)
			return +1;
		if (a.quartal < b.quartal)
			return -1;
		if (a.quartal > b.quartal)
			return +1;
		return 0;
	} };

	/**
	 * Ein Logger für Debug-Zwecke.
	 */
	private readonly _logger : Logger;


	/**
	 * Der Konstruktor.
	 */
	public constructor();

	/**
	 * Der Konstruktor.
	 *
	 * @param pLogger  Ein Logger für Debug-Zwecke.
	 */
	public constructor(pLogger : Logger);

	/**
	 * Implementation for method overloads of 'constructor'
	 */
	public constructor(__param0? : Logger) {
		super();
		if ((typeof __param0 === "undefined")) {
			this._logger = new Logger();
		} else if (((typeof __param0 !== "undefined") && ((__param0 instanceof JavaObject) && ((__param0 as JavaObject).isTranspiledInstanceOf('de.svws_nrw.core.logger.Logger'))))) {
			const pLogger : Logger = cast_de_svws_nrw_core_logger_Logger(__param0);
			this._logger = pLogger;
		} else throw new Error('invalid method overload');
	}

	/**
	 * Berechnet eine Liste von Terminen, denen die IDs der übergebenen Kurs-Klausuren zugeordnetet sind.
	 *
	 * @param daten   die Kurs-Klausuren mit Schülern und die Konfiguration als Eingabe für den Blockungs-Algorithmus
	 *
	 * @return die Liste der Termine
	 */
	public apply(daten : GostKlausurterminblockungDaten) : GostKlausurterminblockungErgebnis {
		const out : GostKlausurterminblockungErgebnis = new GostKlausurterminblockungErgebnis();
		this.berechneRekursivQuartalsModus(daten.klausuren, daten.konfiguration, out);
		return out;
	}

	private berechneRekursivQuartalsModus(input : List<GostKursklausur>, config : GostKlausurterminblockungKonfiguration, out : GostKlausurterminblockungErgebnis) : void {
		if (input.isEmpty())
			return;
		if (config.modusQuartale === KlausurterminblockungModusQuartale.ZUSAMMEN.id) {
			this.berechneRekursivLkGkModus(input, config, out);
			return;
		}
		input.sort(KlausurterminblockungAlgorithmus._compGostKursklausur);
		const temp : List<GostKursklausur> = new ArrayList();
		for (const klausur of input) {
			if (temp.isEmpty()) {
				temp.add(klausur);
				continue;
			}
			if (KlausurterminblockungAlgorithmus._compGostKursklausur.compare(klausur, temp.get(0)) === 0) {
				temp.add(klausur);
				continue;
			}
			this.berechneRekursivLkGkModus(temp, config, out);
			temp.clear();
			temp.add(klausur);
		}
		if (!temp.isEmpty()) {
			this.berechneRekursivLkGkModus(temp, config, out);
			temp.clear();
		}
	}

	private berechneRekursivLkGkModus(input : List<GostKursklausur>, config : GostKlausurterminblockungKonfiguration, out : GostKlausurterminblockungErgebnis) : void {
		const modus : KlausurterminblockungModusKursarten = KlausurterminblockungModusKursarten.getOrException(config.modusKursarten);
		switch (modus) {
			case KlausurterminblockungModusKursarten.BEIDE: {
				this.berechne_helper(input, config, out);
				break;
			}
			case KlausurterminblockungModusKursarten.NUR_LK: {
				this.berechne_helper(KlausurterminblockungAlgorithmus.filter(input, true), config, out);
				break;
			}
			case KlausurterminblockungModusKursarten.NUR_GK: {
				this.berechne_helper(KlausurterminblockungAlgorithmus.filter(input, false), config, out);
				break;
			}
			case KlausurterminblockungModusKursarten.GETRENNT: {
				this.berechne_helper(KlausurterminblockungAlgorithmus.filter(input, true), config, out);
				this.berechne_helper(KlausurterminblockungAlgorithmus.filter(input, false), config, out);
				break;
			}
			default: {
				throw new DeveloperNotificationException("Der LK-GK-Modus ist unbekannt!")
			}
		}
	}

	/**
	 * Liefert die Liste pInput nach LK-Klausuren (oder dem Gegenteil) gefiltert heraus.
	 *
	 * @param input   Die Liste, die gefiltert werden soll.
	 * @param istLK   Falls TRUE, werden die LK-Klausuren herausgefiltert, andernfalls das Gegenteil.
	 *
	 * @return die Liste pInput nach LK-Klausuren (oder dem Gegenteil) gefiltert heraus.
	 */
	private static filter(input : List<GostKursklausur>, istLK : boolean) : List<GostKursklausur> {
		const temp : List<GostKursklausur> = new ArrayList();
		for (const gostKursklausur of input)
			if (JavaObject.equalsTranspiler(gostKursklausur.kursart, ("LK")) === istLK)
				temp.add(gostKursklausur);
		return temp;
	}

	/**
	 * Berechnet eine Klausurblockung unter Verwendung verschiedener internen Algorithmen.
	 *
	 * @param input   die Menge der Klausuren (Eingabe).
	 * @param config  die Konfiguration der Blockung.
	 * @param out     die Termin-Klausur-Zuordnung (Ausgabe).
	 */
	private berechne_helper(input : List<GostKursklausur>, config : GostKlausurterminblockungKonfiguration, out : GostKlausurterminblockungErgebnis) : void {
		this._logger.log("KlausurterminblockungAlgorithmus");
		this._logger.modifyIndent(+4);
		const zeitEndeGesamt : number = System.currentTimeMillis() + config.maxTimeMillis;
		const seed : number = KlausurterminblockungAlgorithmus._random.nextLong();
		const random : Random = new Random(seed);
		const dynDaten : KlausurterminblockungDynDaten | null = new KlausurterminblockungDynDaten(this._logger, random, input, config);
		const algorithmen : Array<KlausurterminblockungAlgorithmusAbstract> = [new KlausurterminblockungAlgorithmusGreedy1(random, dynDaten), new KlausurterminblockungAlgorithmusGreedy1b(random, dynDaten), new KlausurterminblockungAlgorithmusGreedy2(random, dynDaten), new KlausurterminblockungAlgorithmusGreedy2b(random, dynDaten), new KlausurterminblockungAlgorithmusGreedy3(random, dynDaten)];
		dynDaten.aktion_Clear_TermineNacheinander_GruppeZufaellig();
		dynDaten.aktionZustand2Speichern();
		let zeitProAlgorithmus : number = 10;
		do {
			this._logger.log("zeitProAlgorithmus --> " + zeitProAlgorithmus);
			for (let iAlgo : number = 0; iAlgo < algorithmen.length; iAlgo++) {
				const zeitEndeRunde : number = System.currentTimeMillis() + zeitProAlgorithmus;
				algorithmen[iAlgo].berechne(zeitEndeRunde);
				this._logger.log(algorithmen[iAlgo].toString()! + " --> " + dynDaten.gibTerminAnzahl());
			}
			zeitProAlgorithmus *= 2;
		} while (System.currentTimeMillis() + algorithmen.length * zeitProAlgorithmus <= zeitEndeGesamt);
		dynDaten.aktionZustand2Laden();
		out.termine.addAll(dynDaten.gibErzeugeOutput().termine);
		this._logger.modifyIndent(-4);
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.core.utils.klausurplanung.KlausurterminblockungAlgorithmus'].includes(name);
	}

}

export function cast_de_svws_nrw_core_utils_klausurplanung_KlausurterminblockungAlgorithmus(obj : unknown) : KlausurterminblockungAlgorithmus {
	return obj as KlausurterminblockungAlgorithmus;
}

File: svws-core/src/main/java/de/svws_nrw/core/utils/gost/klausurplanung/KlausurterminblockungAlgorithmus.java
package de.svws_nrw.core.utils.gost.klausurplanung;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Random;

import de.svws_nrw.core.data.gost.klausurplanung.GostKursklausurRich;
import de.svws_nrw.core.data.gost.klausurplanung.GostKlausurterminblockungDaten;
import de.svws_nrw.core.data.gost.klausurplanung.GostKlausurterminblockungErgebnis;
import de.svws_nrw.core.data.gost.klausurplanung.GostKlausurterminblockungKonfiguration;
import de.svws_nrw.core.exceptions.DeveloperNotificationException;
import de.svws_nrw.core.logger.Logger;
import de.svws_nrw.core.types.gost.klausurplanung.KlausurterminblockungModusKursarten;
import de.svws_nrw.core.types.gost.klausurplanung.KlausurterminblockungModusQuartale;
import jakarta.validation.constraints.NotNull;

/**
 * Algorithmus zur Blockung von Klausuren auf eine minimale Anzahl von Terminen.
 * Die Termine sind noch keinem Datum zugeordnet.
 *
 * @author Benjamin A. Bartsch
 */
public class KlausurterminblockungAlgorithmus {

	private static final @NotNull Random _random = new Random();

	private static final @NotNull Comparator<@NotNull GostKursklausurRich> _compGostKursklausurRich = (final @NotNull GostKursklausurRich a, final @NotNull GostKursklausurRich b) -> {
		if (a.halbjahr < b.halbjahr) return -1;
		if (a.halbjahr > b.halbjahr) return +1;

		if (a.quartal < b.quartal) return -1;
		if (a.quartal > b.quartal) return +1;

		return 0;
	};


	/** Ein Logger für Debug-Zwecke. */
	private final @NotNull Logger _logger;

	/**
	 * Der Konstruktor.
	 */
	public KlausurterminblockungAlgorithmus() {
		_logger = new Logger();
	}

	/**
	 * Der Konstruktor.
	 *
	 * @param pLogger  Ein Logger für Debug-Zwecke.
	 */
	public KlausurterminblockungAlgorithmus(final @NotNull Logger pLogger) {
		_logger = pLogger;
	}

	/**
	 * Berechnet eine Liste von Terminen, denen die IDs der übergebenen Kurs-Klausuren zugeordnet sind.
	 *
	 * @param daten   die Kurs-Klausuren mit Schülern und die Konfiguration als Eingabe für den Blockungs-Algorithmus
	 *
	 * @return die Liste der Termine
	 */
	public @NotNull GostKlausurterminblockungErgebnis apply(final @NotNull GostKlausurterminblockungDaten daten) {
		final @NotNull GostKlausurterminblockungErgebnis out = new GostKlausurterminblockungErgebnis();
		berechneRekursivQuartalsModus(daten.richKlausuren, daten.konfiguration, out);
		return out;
	}

	private void berechneRekursivQuartalsModus(final @NotNull List<@NotNull GostKursklausurRich> input, final @NotNull GostKlausurterminblockungKonfiguration config, final @NotNull GostKlausurterminblockungErgebnis out) {

		// Keine Daten vorhanden.
		if (input.isEmpty())
			return;

		// Keine Filterung angefordert.
		if (config.modusQuartale == KlausurterminblockungModusQuartale.ZUSAMMEN.id) {
			berechneRekursivLkGkModus(input, config, out);
			return;
		}

		// Sortieren nach (Halbjahr, Quartal).
		input.sort(_compGostKursklausurRich);

		// Blocken nach (Halbjahr, Quartal).
		final @NotNull List<@NotNull GostKursklausurRich> temp = new ArrayList<>();
		for (final @NotNull GostKursklausurRich klausur : input) {
			// Neue Gruppe (Sonderfall) erstellen.
			if (temp.isEmpty()) {
				temp.add(klausur);
				continue;
			}

			// Gleiche Gruppe.
			if (_compGostKursklausurRich.compare(klausur, temp.get(0)) == 0) {
				temp.add(klausur);
				continue;
			}

			// Gruppe blocken
			berechneRekursivLkGkModus(temp, config, out);

			// Neue Gruppe erstellen.
			temp.clear();
			temp.add(klausur);
		}

		if (!temp.isEmpty()) {
			berechneRekursivLkGkModus(temp, config, out);
			temp.clear();
		}
	}



	private void berechneRekursivLkGkModus(final @NotNull List<@NotNull GostKursklausurRich> input, final @NotNull GostKlausurterminblockungKonfiguration config, final @NotNull GostKlausurterminblockungErgebnis out) {
		// Der LK-GK-Modus bestimmt, welche Klausuren aus der Liste potentiell geblockt werden sollen.
		final @NotNull KlausurterminblockungModusKursarten modus = KlausurterminblockungModusKursarten.getOrException(config.modusKursarten);
		switch (modus) {
			case BEIDE:
				berechne_helper(input, config, out); // keine Filterung
				break;
			case NUR_LK:
				berechne_helper(filter(input, true), config, out);  // nur LK
				break;
			case NUR_GK:
				berechne_helper(filter(input, false), config, out);  // nur GK (bzw. nicht LK)
				break;
			case GETRENNT:
				berechne_helper(filter(input, true), config, out);  // nur LK
				berechne_helper(filter(input, false), config, out);  // nur GK (bzw. nicht LK)
				break;
			default:
				throw new DeveloperNotificationException("Der LK-GK-Modus ist unbekannt!");
		}
	}

	/**
	 * Liefert die Liste pInput nach LK-Klausuren (oder dem Gegenteil) gefiltert heraus.
	 *
	 * @param input   Die Liste, die gefiltert werden soll.
	 * @param istLK   Falls TRUE, werden die LK-Klausuren herausgefiltert, andernfalls das Gegenteil.
	 *
	 * @return die Liste pInput nach LK-Klausuren (oder dem Gegenteil) gefiltert heraus.
	 */
	private static @NotNull List<@NotNull GostKursklausurRich> filter(final @NotNull List<@NotNull GostKursklausurRich> input, final boolean istLK) {
		final @NotNull List<@NotNull GostKursklausurRich> temp = new ArrayList<>();

		for (final GostKursklausurRich gostKursklausur : input)
			if (gostKursklausur.kursart.equals("LK") == istLK)
				temp.add(gostKursklausur);

		return temp;
	}

	/**
	 * Berechnet eine Klausurblockung unter Verwendung verschiedener internen Algorithmen.
	 *
	 * @param input   die Menge der Klausuren (Eingabe).
	 * @param config  die Konfiguration der Blockung.
	 * @param out     die Termin-Klausur-Zuordnung (Ausgabe).
	 */
	private void berechne_helper(final @NotNull List<@NotNull GostKursklausurRich> input, final @NotNull GostKlausurterminblockungKonfiguration config, final @NotNull GostKlausurterminblockungErgebnis out) {
		// Logger +4
		_logger.log("KlausurterminblockungAlgorithmus");
		_logger.modifyIndent(+4);

		// End-Zeitpunkt berechnet.
		final long zeitEndeGesamt = System.currentTimeMillis() + config.maxTimeMillis;

		// Random-Objekt erzeugen
		final long seed = _random.nextLong(); // Trick, um den Seed zu kennen.
		final @NotNull Random random = new Random(seed);

		// Erstellung eines Objektes, welches sich um alle dynamischen Daten während des Blockungsvorgangs kümmert.
		final KlausurterminblockungDynDaten dynDaten = new KlausurterminblockungDynDaten(_logger, random, input, config);

		// Algorithmen erzeugen
		final @NotNull KlausurterminblockungAlgorithmusAbstract @NotNull [] algorithmen = new KlausurterminblockungAlgorithmusAbstract @NotNull [] {
				// Alle Algorithmen zur Verteilung von Klausuren auf ihre Termine ...
				new KlausurterminblockungAlgorithmusGreedy1(random, dynDaten), // Klausurgruppen zufällig, Termine zufällig
				new KlausurterminblockungAlgorithmusGreedy1b(random, dynDaten), // Termine nacheinander, Klausurgruppen zufällig
				new KlausurterminblockungAlgorithmusGreedy2(random, dynDaten), // Klausurgruppen nach Grad, Termine zufällig
				new KlausurterminblockungAlgorithmusGreedy2b(random, dynDaten), // Termine nacheinander, Klausurgruppen nach Grad
				new KlausurterminblockungAlgorithmusGreedy3(random, dynDaten), // Termine nacheinander, Klausurgruppen mit Backtracking
				// ... Ende der Algorithmen.
		};


		// Generiere das erste Ergebnis und speichere es in Zustand2 (bisheriges globales Optimum).
		dynDaten.aktion_Clear_TermineNacheinander_GruppeZufaellig();
		dynDaten.aktionZustand2Speichern();

		long zeitProAlgorithmus = 10L; // Weniger ist nicht gut.
		do {
			_logger.log("zeitProAlgorithmus --> " + zeitProAlgorithmus);

			// Alle Algorithmus starten (pro Algorithmus ggf. auch mehrfach).
			for (int iAlgo = 0; iAlgo < algorithmen.length; iAlgo++) {
				final long zeitEndeRunde = System.currentTimeMillis() + zeitProAlgorithmus;
				algorithmen[iAlgo].berechne(zeitEndeRunde);
				_logger.log(algorithmen[iAlgo].toString() + " --> " + dynDaten.gibTerminAnzahl());
			}

			zeitProAlgorithmus *= 2; // Nächste Runde hat mehr Zeit.
		} while (System.currentTimeMillis() + algorithmen.length * zeitProAlgorithmus <= zeitEndeGesamt); // noch Zeit?

		// Lade besten globalen Zustand
		dynDaten.aktionZustand2Laden();

		// Fülle die Liste mit Termin-Listen.
		out.termine.addAll(dynDaten.gibErzeugeOutput().termine);

		// Logger -4
		_logger.modifyIndent(-4);
	}

}
